### PR TITLE
[GUI] Major dependency upgrades

### PR DESCRIPTION
**Pull request by BenjaminBrandmeier**
### Major dependency upgrades (Electron + Web)

The following changes have been made:
- Angular 7 to 8
- Nebular 3 to 4
- Upgrading all related dependencies
- Code changes to preserve the old look and feel

Those changes are pre-requirements for the following long term goals:
- Node 16 support
- Multiple theme support (like an additional dark mode https://github.com/TestRunnerSRL/OoT-Randomizer/issues/1380)

The long term goals are not part of this PR but I might be willing to work on this, if this PR gets merged. 
I have noticed that there have been plenty of questions in Discord about broken installations, because the current Node LTS version is not supported. According to the following table, we should get there with Angular 13 latest https://gist.github.com/LayZeeDK/c822cc812f75bb07b7c55d07ba2719b3

### Implementation details:

**No more View Encapsulation** 

Since bootstrap is used by the generator as well as the website, Shadow-DOM was introduced to not mess with the websites styling (and vice-versa). When upgrading to Nebular 4.6 I have noticed some issues already fixed in a private fork by @dragonbane0 related to the usage of Shadow-DOM (described here: https://github.com/akveo/nebular/issues/1652). Since nebular still does not have support for Shadow-DOM, I have migrated those fixes to a separate 4.6 fork at first, but ultimately decided against that approach since it causes some weird dependency errors when distributing the adjusted 4.6 version. To ensure the website does not break I have simply adjusted some styles and made sure that the webTest environment behaves the same as before.

**Stylesheet adjustments**

Migrating to Nebular 4 required some major changes described in the 3.x->4 Nebular upgrade guide. After the migration the look and feel was a bit of a mess, so I have made adjustments to keep it as we have it currently. I thought you are not too keen about updating all screenshots and video tutorials. ;-)

It's not a 100% pixel perfect, but if you are not taking a very precise look at the now and then version you probably won't even notice - at least that was the goal. While doing these backwards compatibility adjustments I have removed plenty of unused stylesheets. 

To be able to switch between themes in the future easily we should use theme variables whenever any type of colors (background, border, font, etc.) are specified within the code. Like `color: nb-theme(color-basic-100);` for example.

### Tests

I have tested the installation process (Gui.py, npm i, npm run ng-dev, npm run ng-release, etc.) for the Electron app as well as the webTest environment on Linux and Windows with Firefox and Chrome. Since those changes were only GUI changes there shouldn't be any side effects to the ROM manipulation itself. At least I have ran one just fine.

As with my latest PR the tag **needs testing** should probably be added to this issue to ensure full compatibility.

### CODE REVIEW

**Comment by TreZc0** [comment]
Wow outstanding. We have not expected anyone to give this a shot before we were able to tackle it ourselves.

Thank you so much for your work already, I will give this a thorough check soon and probably have some feedback, but am also willing to join in getting the next few steps done.

**Comment by ETR-BTF** [comment]
On the Main Rules tab I cannot enter numbers for the Rainbow Bridge or Trials. These fields work fine on the Other tab tho.

**Comment by ETR-BTF** [comment]
Also, I'd love to see your Dark Mode theme. Is there like a change in the code I can make to unlock that? Since it doesn't seem to have its own GUI setting/toggle yet as far as I can tell.

**Comment by BenjaminBrandmeier** [comment]
@TreZc0 Thanks a lot. I hope the integration with the webpage works smoothly. Looking forward to your feedback.

@ETR-BTF 
For me the Rainbow Bridge and Trials are working fine (Electron + Web). Can you give me some details about your setup + browser/OS you are testing this on?

Oh, and regarding the dark theme. I haven't built it yet, but after those code changes it should be easier to create one since we are using the "new" (in fact still 2019) Nebular 4 theming approach.

**Comment by ETR-BTF** [comment]
I'm not using a browser I'm just running Gui.py. I'm still using Node v10.16.2 (lol) so maybe that's why it doesn't work for me. Still, it's odd that it only doesn't work on that tab.

image

(@ dark mode) Oh, I see, I misunderstood what that part of the PR does then.

**Comment by BenjaminBrandmeier** [comment]
@ETR-BTF should be fixed. This only happened for specific resolutions.

**Comment by ETR-BTF** [comment]
Ah, gotcha. I can confirm it works on my end too, thanks.

**Comment by TreZc0** [comment]
As I feared, a few things on the website don't take lightly that you mostly removed the capsulation on real web.
image
The nav goes vertical and the instructions button instantly closes the panel below again.

These are small issues tho and issues that can be fixed, some even on web instead of the rando itself (which means a lot less hassle...).

During this, I noticed that the webtest page should get a few updates to properly reflect the website in its current state, which is something I will do soon (on vacation rn, so everything is a little slow).

As for testing, I could also offer you a way to deploy your fork and branch directly to the website for more testing, if that's something you are interested in.

In general tho: Fantastic work. I checked the GUI look yesterday, and the small changes, especially in regards to the font in some places, feel very good and cleaner. I like it.

**Comment by BenjaminBrandmeier** [comment]
That's kind of what I have expected as well. 
The only options I've had regarding web compatibility was working with the provided files in the webTest directory.

So yeah, any way to move further with this (updated webTest and/or fork integration) would be great. Looking at your screenshot I assume there will be a few changes to the websites code (in addition to the small ones I have made to [generator5Page.css](https://github.com/BenjaminBrandmeier/OoT-Randomizer/blob/gui-upgrade/GUI/webTest/public/css/generator5Page.css) and [mediaCSS5.css](https://github.com/BenjaminBrandmeier/OoT-Randomizer/blob/gui-upgrade/GUI/webTest/public/css/mediaCSS5.css) for now.

Enjoy your vacation! 
I'm not impatient about this PR and will also have to take my time for making the adjustments. In addition I assume the upcoming upgrades up to Angular 13 with Node 16 support will be much easier once we have solved the remaining issues in this PR.

**Comment by BenjaminBrandmeier** [comment]
I have further enhanced the style encapsulation (without ShadowDom). As I've thought Instruction/Help panel can easily be fixed by adding four lines to the (outdated) generator5Page.css. 

For the current webTest it's working now. However, I will wait with further updates on this PR until TreZc0 has updated the webTest directory to be in line with the current state of the website and then migrate my modifications to that.

**Comment by cjohnson57** [comment]
Just going to echo the sentiment that you've done amazing work on this although I will leave the reviewing of the PR to TreZ/db0

**Comment by TreZc0** [comment]
#1731 will take care of the webtest update. Will hand out a deployment pipeline for you to test on web later :)

**Comment by BenjaminBrandmeier** [comment]
Thanks TreZ! By granting me access to the dev portal and the updated webtest directory it was pretty straight-forward to resolve the remaining website compatibility issues. 

@cjohnson57 Thanks a lot! Once https://github.com/TestRunnerSRL/OoT-Randomizer/pull/1731 gets merged to Dev I will push the remaining code changes to this PR.

**Comment by BenjaminBrandmeier** [comment]
Squashed commits for better readability and reverted all changes done within the webTest directory. (no changes to the webpage itself are required)

Considering this PR ready to merge.

**Comment by BenjaminBrandmeier** [comment]
Just found another tiny thing that was fixed by the force-push above.


before: before
after: after

Please be aware of the changed version before merging, those changes are required to work with the website test integration.

**Comment by Cuphat** [comment]
Noticed a few small issues on the electron version.

image

The WAD Channel ID and WAD Channel Title text boxes are too short and the Multi-World and Settings Presets panels have a bit too much padding at the top which makes them overflow a bit. For reference, here's the same section on the current GUI

image

File: version.py [hunk 1]
```diff
@@ -1,4 +1,4 @@
-__version__ = '6.2.216'
+__version__ = '6.2.255'
```
Comment (mracsys) [inline version.py]: Is there a reason this was bumped to 255? Caught me off guard merging this into my branch.

Comment (BenjaminBrandmeier) [reply version.py]: I had to update the version to something unique to appropriately work with the pipeline for the website test integration from TreZ. Not sure what the best practice is here in general, but for future PRs I think I will change this on identical duplicate branches for website testing.

So please feel free to change this to something appropriate when merging.

**Comment by BenjaminBrandmeier** [comment]
@Cuphat Thanks for the feedback. You are absolutely right. One of those were actually fixed in the upcoming dark mode branch which I haven't pushed yet. I will push these adjustments on this branch as soon as I find some time within the next days.

**Comment by mracsys** [comment]
Another minor cosmetic issue on the Cosmetics tab (ha). The "Outer" description text overlaps the inner dropdowns at certain widths.
image

**Comment by BenjaminBrandmeier** [comment]
Reset version and provided fixes.

**WAD + Overflow:**

cuphat_fixes

**Colors overlap:**

fix_mracsys

File: version.py (hunk 1 above)
Comment (BenjaminBrandmeier) [reply version.py]: With the latest commit I have reset the version to 6.2.216

**Comment by mracsys** [comment]
I got a report that DPI scaling on Windows 10 made some controls inaccessible. User had the following settings:
* Resolution: 1920x1080
* Scaling: 125%

Controls at the bottom of scrollable fields would not respond to clicks, such as the Load/Save/Remove buttons for presets. Randomize Overworld Spawns was another problem on the Main Rules tab. Screenshot from the user with the GUI fullscreen:

image

**Comment by dragonbane0** [comment]
Does this work on older versions? DPI scaling is a huge issue for many apps

**Comment by mracsys** [comment]
They did not have issues on versions with the old GUI.

**Comment by dragonbane0** [comment]
Alright. Good to know about it then

**Comment by TreZc0** [comment]
This below is probably a good way to go about it for now, adding support with our dynamic sections sounds painful.

https://github.com/electron/electron/issues/6571#issuecomment-326865765

**Comment by BenjaminBrandmeier** [comment]
I have tested with 125% on Windows and can confirm the issue and will provide a fix for this. Will try to tweak some values and only disabled DPI scaling as a last resort.

@mracsys the fixes will also be pushed on the dark-mode branch (since you have found that one already) ;)

**Comment by dragonbane0** [comment]
If it did work fine with the old version, disabling DPI scaling would be a regression, so should definetly be avoided. But sounds like you are already working on it

**Comment by BenjaminBrandmeier** [comment]
I agree with db0, we should avoid regression.

The changes made fix all the mentioned issues without disabling DPI scaling. In addition there is more space now on the website for configurations.

Kindly recheck.

**Comment by BenjaminBrandmeier** [comment]
rebased and reworked some hint alignments.

**Comment by dragonbane0** [comment]
Looking great! Did some throughout testing now and found some more bugs to report:

1. Textboxes in Prompts should be full length again. [Old](https://i.imgur.com/G8Ub9nY.png), [New](https://i.imgur.com/ThRYszJ.png)

2. Yes/No prompts are missing the Yes button, maybe scale the buttons back to their original size for usability as well. [Bug](https://i.imgur.com/PT2iGSI.png)

3. Trackbars have a very large mouse capture area where clicking vastly outside of the element will still affect the slider. This used to be an issue with the old GUI as well, but was fixed. Example of click in new GUI that affects the slider, but shouldn't: [Example](https://i.imgur.com/QW4BMue.png)

4. Slightly broken border here in Select Boxes: [Click](https://i.imgur.com/WmjS5ri.png)

5. Responsiveness has seen some downgrades. Using Pixel 5 size, the old GUI looked like this: [Click](https://i.imgur.com/IfKSXy3.png), while the new GUI will look like this: [Click](https://i.imgur.com/WfKsI90.png). The old GUI scaled the text down more and broke lines to use space optimally. The bottom section sees a similar issue where the old GUI could fit Copy/Get and Import onto the same line, while the new GUI will break the line like this: [Click](https://i.imgur.com/FX5d045.png)

6. When you generate a local seed and cause an error (easily done intentionally by disabling too many locations) and hit the Details button, the extra error details page has a very small Copy button, that one should be scaled to have the same height as the OK button, also they used to have a distinct color from each other. [Click](https://cdn.discordapp.com/attachments/167048669198155778/1034212138341908530/unknown.png)

Apart from that the rest looks excellent to me. Thanks for the great PR!

**Comment by BenjaminBrandmeier** [comment]
Thanks for the thorough testing and the detailed screenshots db0! 

Few things were already fixed in the upcoming Angular 13 + Dark mode branch. Others I haven't noticed so far, but should all be doable.

Will provide fixes here and cherry-pick to to the upcoming branch as well.

**Comment by BenjaminBrandmeier** [comment]
The last commit should fix all of the reported issues. Please recheck, thanks!

**Review by cjohnson57 — APPROVED** [review]
Great job on this, seems good to merge

**Pull request merged**